linux_os/guide/system/permissions/files/permissions_within_important_dirs/file_ownership_library_dirs/rule.yml

```diff
@@ -69,5 +69,6 @@ template:
             - /lib64/
             - /usr/lib/
             - /usr/lib64/
+        recursive: 'true'
```

Comment:
is this really needed or is it an issue related to PR #8194 ?

Reply:
It is necessary, the check is not going down the directory tree.
Check this test scenario: https://github.com/ComplianceAsCode/content/blob/d5a23b41fad45249f125258225737f96300113fb/linux_os/guide/system/permissions/files/permissions_within_important_dirs/file_ownership_library_dirs/tests/incorrect_owner_within_dir.fail.sh

Reply:
I think we are looking at the same problem, but proposed different approaches.

Reply:
maybe I'm confused (I'll be honest, it's some many rules touching permissions that I have to re-read it all the time), but this rule specifically has the file_regex, shouldn't it be working recursively as you added to the documentation?
I ran a test and can see that the bash (not sure on ansible though) is correctly going through all the files. Isn't it just a matter of fixing the oval?

Reply:
> but this rule specifically has the file_regex, shouldn't it be working recursively as you added to the documentation?

Hmm, my addition in the documentation is about what will be checked, directories vs files. It doesn't address recursion.
a. If `filepath` is a directory and `file_regex` is not specified, the rule will check and remediate the first level of directories in the `filepath`.
b. If `filepath` is a directory and `file_regex` is specified, the rule will check and remediate files in the `filepath`.

If along with case `a.`, recursion is set, the rule will check directories while recursing down.
If along with case `b.`, recursion is set, the rule will check files while recursing down.

> I ran a test and can see that the bash (not sure on ansible though) is correctly going through all the files.

Yes, the remediation is going down the directory tree, but should it happen when I didn't set `recursive`?

> Isn't it just a matter of fixing the oval?

Yes, the OVAL was the part that initially needed fixing, but then I addressed inconsistencies between what the OVAL does and the remediation does.
I think if the check recurses, the remediation should too. But if the OVAL doesn't recurse, the remediations shouldn't recurse.

Reply:
> Yes, the remediation is going down the directory tree, but should it happen when I didn't set `recursive`?

hmm got your point now.

I believe we have more rules that will need the recursive flag then, besides the two rules you touched.
I actually need to go through Ubuntu 20.04 STIG and check all the rules and what is needed on which rule.

I also believe that PR #8194 is still needed because we are not dealing with the fact that a filepath might be a symlink

Reply:
I think we should have more examples in the template documentation, but that should be done in another PR.

Reply:
> I also believe that PR #8194 is still needed because we are not dealing with the fact that a filepath might be a symlink

Yeah, in RHEL `/lib` is a symlink to `/usr/lib`. 
@dodys Do you mean when the symlink itself is missing, or when the path pointed to by the symlink is missing?
It seems to me that `missing_file_pass` will affect the first case, when the symlink itself is absent.
I think [symlink_test](https://oval-community-guidelines.readthedocs.io/en/latest/oval-schema-documentation/unix-definitions-schema.html?highlight=recurse_direction#symlink-test) will be needed to check if the file pointed to is missing.

Out of a review item with Matthew I filed https://github.com/ComplianceAsCode/content/issues/8412.
The expected behaviour of the round around symlinks is not clear to me.

Reply:
> Yeah, in RHEL `/lib` is a symlink to `/usr/lib`. @dodys Do you mean when the symlink itself is missing, or when the path pointed to by the symlink is missing? It seems to me that `missing_file_pass` will affect the first case, when the symlink itself is absent. I think [symlink_test](https://oval-community-guidelines.readthedocs.io/en/latest/oval-schema-documentation/unix-definitions-schema.html?highlight=recurse_direction#symlink-test) will be needed to check if the file pointed to is missing.
> 

No, my point is that with the mentioned commit id, it started to filter out symlinks in the oval. Therefore, if `/lib` is a symlink it won't be checked. If it won't be checked than the number of files evaluated is 0 for that test, thus evaluating to false and the whole rule will fail as the oval is set to all_exist. By passing `missing_file_pass` in the rule, then it changes from `all_exist` to `any_exist` and allow the rule to pass. It is a quick fix. 

> Out of a review item with Matthew I filed #8412. The expected behaviour of the rule around around symlinks is not clear to me.

I don't really remember if most of the permissions/ownership/group ownership rules really mention about symlinks. If oval is set to exclude symlinks (as it is right now), then I don't see the point (and feel free to correct me here) of passing -h in the remediation.

Reply:
> Therefore, if `/lib` is a symlink it won't be checked. If it won't be checked than the number of files evaluated is 0 for that test, thus evaluating to `false` and the whole rule will fail as the oval is set to `all_exist`.

From what I could test with changes in this PR  (using `--oval-results`), in RHEL, even when `/lib` is a symlink the scanner was able to collect objects.
So actually, could it be that the scanner follows symlinks, but doesn't report them as collected objects, :thinking: ?

Attached is the oval-results file:
[rhel7-oval-results.zip](https://github.com/ComplianceAsCode/content/files/8340665/rhel7-oval-results.zip)

Reply:
> I don't really remember if most/any of the permissions/ownership/group ownership rules really mention about symlinks. 

I don't think symlinks are mentioned anywhere. 

> If oval is set to exclude symlinks (as it is right now), then I don't see the point (and feel free to correct me here) of passing -h in the remediation.

I see, it is strange to change the owner of the symlink when the check ignores them.
The intent of the change was to avoid dereferencing erors when the symlink is broken. (I made that commit when I was still not aware of the symlink mess)
With `-h` it means that `chown` will not affect the file pointed to, but as the the `find` command doesn't follow symlinks by default ([and `-L` is not set](https://github.com/ComplianceAsCode/content/pull/8404/files#diff-dd85d2fec512415762ebc6bd5d4f47792cea008a4b21940a20514ea3b6a87448R16)), the file pointed to by the symlink wouldn't be changed anyway...but now we avoid dereferencing errors!

And I see inconsistency [here](https://github.com/ComplianceAsCode/content/pull/8404/files#diff-dd85d2fec512415762ebc6bd5d4f47792cea008a4b21940a20514ea3b6a87448R23) too.

To better align with the check the remediation should filter out symlinks before doing `chown`... But then we are not sure if we actually want to filter out the symlinks...

I'm aware that bdc59897f25e4541d9e9bd2db5eb13ef11702252 doesn't fix the symlinking situation, it just avoids errors in `chwon` command with the current behaviour.
This PR was to improve inconsistency with `recurse` and `file_regex` and I stumbled upon `symlink` and `chwon` inconsistency, :joy:

Reply:
I've just checked here my test from yesterday on master,
so `/lib` and `/lib64` are both symlinks to `/usr/lib` and `/usr/lib64` respectively.
One of the rules that I see failing `file_ownership_library_dirs` will fail for `/lib64` as the only file inside it is a .so that is a symlink to a `.so` under `/usr/lib64`.
It doesn't fail for `/lib` as it "finds" some files there, which are actually files under `/usr/lib`.

It is tricky and I need to dig deeper on those rules, just not having the time lately.
I have to test with your changes, it might be that all this recursive and changes might fix some of the weirdness I saw in the failing tests.

Reply:
Yes, it is tricky. I had to take some time to wrap my head around what the rules do.

Let us know how the tests on your side go.